### PR TITLE
Add unit tests for StringUtils and TimeUtils

### PR DESCRIPTION
**Pull request by shafi-1**
I've analysed your codebase and noticed that `azkaban-common/src/test/java/azkaban/utils/StringUtils` and `az-core/src/test/java/azkaban/utils/TimeUtils` are not fully tested.
I've written some tests for the methods in these classes with the help of Diffblue Cover.

Hopefully, these tests will help you detect any regressions caused by future code changes. If you would find it useful to have additional tests written for this repository, I would be more than happy to look at other classes that you consider important in a subsequent PR.

### CODE REVIEW

**Comment by paulkennethkent** [comment]
@shafi-1 You need to clean up the commit history

File: az-core/src/test/java/azkaban/utils/TimeUtilsTest.java [hunk 1]
```diff
@@ -0,0 +1,105 @@
+package azkaban.utils;
```
Comment (Braavos6) [inline az-core/src/test/java/azkaban/utils/TimeUtilsTest.java:17]: Header missing

File: azkaban-common/src/test/java/azkaban/utils/StringUtilsTest.java [hunk 2]
```diff
@@ -3,78 +3,138 @@
 import org.junit.Assert;
```
Comment (Braavos6) [inline azkaban-common/src/test/java/azkaban/utils/StringUtilsTest.java:19]: Header missing

File: az-core/src/test/java/azkaban/utils/TimeUtilsTest.java [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+import org.junit.rules.ExpectedException;
+
+public class TimeUtilsTest {
+
+    @Rule public final ExpectedException thrown = ExpectedException.none();
+
+    @Test
+    public void testFormatDateTime() {
+        Assert.assertEquals("-",
+                TimeUtils.formatDateTime(-8_070_450_532_247_928_832L));
```
Comment (Braavos6) [inline az-core/src/test/java/azkaban/utils/TimeUtilsTest.java]: Add a return type which is an actual date

File: az-core/src/test/java/azkaban/utils/TimeUtilsTest.java [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+    }
+
+    @Test
+    public void testFormatDateTimeZone() {
+        Assert.assertEquals("-",
+                TimeUtils.formatDateTimeZone(-8_070_450_532_247_928_832L));
```
Comment (Braavos6) [inline az-core/src/test/java/azkaban/utils/TimeUtilsTest.java]: Add a return type which is an actual date

File: az-core/src/test/java/azkaban/utils/TimeUtilsTest.java [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+    }
+
+    @Test
+    public void testFormatDuration() {
```
Comment (Braavos6) [inline az-core/src/test/java/azkaban/utils/TimeUtilsTest.java]: branch coverage for
`    if (endTime == -1) {
      durationMS = System.currentTimeMillis() - startTime;`

is missing

File: az-core/src/test/java/azkaban/utils/TimeUtilsTest.java [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+        Assert.assertEquals("60 minute(s)",
+                TimeUtils.formatPeriod(TimeUtils.parsePeriodString("60m")));
+        Assert.assertEquals("12 month(s)",
+                TimeUtils.formatPeriod(TimeUtils.parsePeriodString("12M")));
+        Assert.assertEquals("120 second(s)",
+                TimeUtils.formatPeriod(TimeUtils.parsePeriodString("120s")));
+    }
+
+    @Test
+    public void testCreatePeriodString() {
```
Comment (Braavos6) [inline az-core/src/test/java/azkaban/utils/TimeUtilsTest.java]: branch and line coverage add test case with parameter string `null` and `n`

File: az-core/src/test/java/azkaban/utils/TimeUtilsTest.java [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+        Assert.assertEquals("Days",
+                TimeUtils.parsePeriodString("2d").getPeriodType().getName());
+        Assert.assertEquals("Hours",
+                TimeUtils.parsePeriodString("24h").getPeriodType().getName());
+        Assert.assertEquals("Minutes",
+                TimeUtils.parsePeriodString("60m").getPeriodType().getName());
+        Assert.assertEquals("Months",
+                TimeUtils.parsePeriodString("12M").getPeriodType().getName());
+        Assert.assertEquals("Seconds",
+                TimeUtils.parsePeriodString("120s").getPeriodType().getName());
```
Comment (Braavos6) [inline az-core/src/test/java/azkaban/utils/TimeUtilsTest.java]: Add one test case for `IllegalArgumentException`

File: az-core/src/test/java/azkaban/utils/TimeUtilsTest.java [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+    }
+}
```
Comment (Braavos6) [inline az-core/src/test/java/azkaban/utils/TimeUtilsTest.java]: Add blank line

File: az-core/src/test/java/azkaban/utils/TimeUtilsTest.java [hunk 9]
```diff
@@ -0,0 +1,105 @@
...
+        Assert.assertEquals("60m", TimeUtils.
+                createPeriodString(TimeUtils.parsePeriodString("60m")));
+        Assert.assertEquals("12M", TimeUtils.
+                createPeriodString(TimeUtils.parsePeriodString("12M")));
+        Assert.assertEquals("120s", TimeUtils.
+                createPeriodString(TimeUtils.parsePeriodString("120s")));
+    }
+
+    @Test
+    public void testParsePeriodString() throws NullPointerException {
```
Comment (Braavos6) [inline az-core/src/test/java/azkaban/utils/TimeUtilsTest.java]: Move this method one higher.. Above this `testCreatePeriodString`

File: azkaban-common/src/test/java/azkaban/utils/StringUtilsTest.java [hunk 10]
```diff
@@ -3,78 +3,138 @@
...
+
+  @Test
+  public void testShellQuote() {
+	Assert.assertEquals("!\\!!", StringUtils.shellQuote("!", '!'));
+	Assert.assertEquals("\u0000!\u0000",
+			StringUtils.shellQuote("!", '\u0000'));
+  }
+
+  @Test
+  public void testJoinArrayListNull() {
```
Comment (Braavos6) [inline azkaban-common/src/test/java/azkaban/utils/StringUtilsTest.java]: why didn't you put all the tests for `join` in one method?

File: azkaban-common/src/test/java/azkaban/utils/StringUtilsTest.java [hunk 11]
```diff
@@ -3,78 +3,138 @@
...
+	Assert.assertEquals("", StringUtils.join(new ArrayList<>(), null));
+  }
+
+  @Test
+  public void testJoinInputCollectionString() {
+	final Collection<String> list = new ArrayList<>();
+	list.add("!!");
+	list.add("??");
+
+	Assert.assertEquals("!!$$$??$$$", StringUtils.join(list, "$$$"));
```
Comment (Braavos6) [inline azkaban-common/src/test/java/azkaban/utils/StringUtilsTest.java]: you could use Arrays.asList() to do 
`	Assert.assertEquals("!!$$$??$$$",
            StringUtils.join(new ArrayList<>(Arrays.asList("!!","??")),
                    "$$$"));`

File: azkaban-common/src/test/java/azkaban/utils/StringUtilsTest.java [hunk 12]
```diff
@@ -3,78 +3,138 @@
...
+  }
+
+  @Test
+  public void testJoinArrayListString() {
+	final ArrayList<String> list = new ArrayList<>();
+	list.add("");
+
+	Assert.assertEquals("!", StringUtils.join(list, "!"));
```
Comment (Braavos6) [inline azkaban-common/src/test/java/azkaban/utils/StringUtilsTest.java]: same here and below

File: azkaban-common/src/test/java/azkaban/utils/StringUtilsTest.java [hunk 13]
```diff
@@ -3,78 +3,138 @@
...
+	list.add("!!");
+	list.add("??");
+
+	Assert.assertEquals(
+			"!!$$$??", StringUtils.join2(list, "$$$"));
+  }
+
+  @Test
+  public void testIsFromBrowser() {
+	Assert.assertFalse(StringUtils.isFromBrowser(null));
```
Comment (Braavos6) [inline azkaban-common/src/test/java/azkaban/utils/StringUtilsTest.java]: Add these for line and branch coverage
`      Assert.assertFalse(StringUtils.isFromBrowser("foo"));
      Assert.assertTrue(StringUtils.isFromBrowser("*Gecko"));
      Assert.assertTrue(StringUtils.isFromBrowser("*Chrome"));
      Assert.assertTrue(StringUtils.isFromBrowser("*Trident"));
      Assert.assertTrue(StringUtils.isFromBrowser("*AppleWebKit"));`

**Review by Braavos6 — CHANGES_REQUESTED** [review]
Check the spacing for this project please

File: az-core/src/test/java/azkaban/utils/TimeUtilsTest.java (hunk 1 above)
Comment (paulkennethkent) [reply az-core/src/test/java/azkaban/utils/TimeUtilsTest.java:17]: Licence header

File: az-core/src/test/java/azkaban/utils/TimeUtilsTest.java (hunk 3 above)
Comment (paulkennethkent) [reply az-core/src/test/java/azkaban/utils/TimeUtilsTest.java]: What Eric means is that you want to pass in something that asserts a real date

File: az-core/src/test/java/azkaban/utils/TimeUtilsTest.java (hunk 4 above)
Comment (paulkennethkent) [reply az-core/src/test/java/azkaban/utils/TimeUtilsTest.java]: What Eric means is that you want to pass in something that asserts a real date

File: az-core/src/test/java/azkaban/utils/TimeUtilsTest.java (hunk 5 above)
Comment (shafi-1) [reply az-core/src/test/java/azkaban/utils/TimeUtilsTest.java]: I tried doing this, but I'd have to use mocking and it's difficult building the project when powermockito is added as it conflicts with the existing mockito dependencies

File: az-core/src/test/java/azkaban/utils/TimeUtilsTest.java (hunk 9 above)
Comment (shafi-1) [reply az-core/src/test/java/azkaban/utils/TimeUtilsTest.java]: done

File: azkaban-common/src/test/java/azkaban/utils/StringUtilsTest.java (hunk 10 above)
Comment (shafi-1) [reply azkaban-common/src/test/java/azkaban/utils/StringUtilsTest.java]: done

File: azkaban-common/src/test/java/azkaban/utils/StringUtilsTest.java (hunk 11 above)
Comment (shafi-1) [reply azkaban-common/src/test/java/azkaban/utils/StringUtilsTest.java]: changed

**Review by paulkennethkent — CHANGES_REQUESTED** [review]

File: azkaban-common/src/test/java/azkaban/utils/StringUtilsTest.java (hunk 12 above)
Comment (shafi-1) [reply azkaban-common/src/test/java/azkaban/utils/StringUtilsTest.java]: changed except one of them where it had to be a Collection to obtain higher coverage

**Comment by paulkennethkent** [comment]
I think most of what has Eric makes sense. If you don't understand certain parts then I can sit down and help.

File: azkaban-common/src/test/java/azkaban/utils/StringUtilsTest.java (hunk 13 above)
Comment (shafi-1) [reply azkaban-common/src/test/java/azkaban/utils/StringUtilsTest.java]: added foo but the rest oddly break the test

File: az-core/src/test/java/azkaban/utils/TimeUtilsTest.java (hunk 1 above)
Comment (shafi-1) [reply az-core/src/test/java/azkaban/utils/TimeUtilsTest.java:17]: added license header

File: azkaban-common/src/test/java/azkaban/utils/StringUtilsTest.java (hunk 2 above)
Comment (shafi-1) [reply azkaban-common/src/test/java/azkaban/utils/StringUtilsTest.java:19]: added license header

File: az-core/src/test/java/azkaban/utils/TimeUtilsTest.java (hunk 3 above)
Comment (shafi-1) [reply az-core/src/test/java/azkaban/utils/TimeUtilsTest.java]: returns an actual date now, kept the previous assertion for added coverage

File: az-core/src/test/java/azkaban/utils/TimeUtilsTest.java (hunk 4 above)
Comment (shafi-1) [reply az-core/src/test/java/azkaban/utils/TimeUtilsTest.java]: returns an actual date

File: az-core/src/test/java/azkaban/utils/TimeUtilsTest.java (hunk 1 above)
Comment (petersear) [reply az-core/src/test/java/azkaban/utils/TimeUtilsTest.java:17]: I can't see the licence header. On a quick check, I didn't see some of the other changes either.
Is there a new commit which hasn't been pushed yet?

Comment (shafi-1) [reply az-core/src/test/java/azkaban/utils/TimeUtilsTest.java:17]: i'll check

Comment (shafi-1) [reply az-core/src/test/java/azkaban/utils/TimeUtilsTest.java:17]: should be updated now :)

File: az-core/src/test/java/azkaban/utils/TimeUtilsTest.java [hunk 14]
```diff
@@ -0,0 +1,129 @@
...
+	public void testFormatDuration() {
+		Assert.assertEquals("-",
+				TimeUtils.formatDuration(-1L, -513L));
+		Assert.assertEquals("1m 1s",
+				TimeUtils.formatDuration(-65793L, -4609L));
+		Assert.assertEquals("1h 14m 1s",
+				TimeUtils.formatDuration(-4462057L, -20994L));
+		Assert.assertEquals("1290d 15h 0m",
+				TimeUtils.formatDuration(-833737520625L, -722227520546L));
+		Assert.assertEquals("7 sec",
```
Comment (petersear) [inline az-core/src/test/java/azkaban/utils/TimeUtilsTest.java]: The duration increases until the last assert. I think it would sense for this assert to appear above the "1m 1s" assert.
Also, the long literals here contain underscores whereas none of the others do - make these consistent.

File: az-core/src/test/java/azkaban/utils/TimeUtilsTest.java [hunk 15]
```diff
@@ -0,0 +1,129 @@
...
+	}
+
+	@Test
+	public void testParsePeriodStringThrowsException() {
+		thrown.expect(IllegalArgumentException.class);
+		TimeUtils.parsePeriodString("1234");
+	}
+
+	@Test
+	public void testFormatPeriod() {
```
Comment (petersear) [inline az-core/src/test/java/azkaban/utils/TimeUtilsTest.java]: The asserts in `testFormatPeriod()` should be ordered logically. I would suggest testing in the order the method under test checks the time period ie. null first then years, months, weeks, day, hours, minutes, seconds.
I think if you add a test with period of `0`, you will cover the last `return` in the method.

File: az-core/src/test/java/azkaban/utils/TimeUtilsTest.java [hunk 16]
```diff
@@ -0,0 +1,129 @@
...
+		Assert.assertEquals("60 minute(s)",
+				TimeUtils.formatPeriod(TimeUtils.parsePeriodString("60m")));
+		Assert.assertEquals("12 month(s)",
+				TimeUtils.formatPeriod(TimeUtils.parsePeriodString("12M")));
+		Assert.assertEquals("120 second(s)",
+				TimeUtils.formatPeriod(TimeUtils.parsePeriodString("120s")));
+	}
+
+	@Test
+	public void testParsePeriodString() throws NullPointerException {
```
Comment (petersear) [inline az-core/src/test/java/azkaban/utils/TimeUtilsTest.java]: List the asserts in a logical order - probably the order they are checked in the method.

File: az-core/src/test/java/azkaban/utils/TimeUtilsTest.java [hunk 17]
```diff
@@ -0,0 +1,129 @@
...
+		Assert.assertEquals("60m", TimeUtils.
+				createPeriodString(TimeUtils.parsePeriodString("60m")));
+		Assert.assertEquals("12M", TimeUtils.
+				createPeriodString(TimeUtils.parsePeriodString("12M")));
+		Assert.assertEquals("120s", TimeUtils.
+				createPeriodString(TimeUtils.parsePeriodString("120s")));
+	}
+
+	@Test
+	public void testParsePeriodStringException() throws IllegalArgumentException {
```
Comment (petersear) [inline az-core/src/test/java/azkaban/utils/TimeUtilsTest.java]: This test should follow testParsePeriodString() rather than testCreatePeriodString()

File: az-core/src/test/java/azkaban/utils/TimeUtilsTest.java [hunk 18]
```diff
@@ -0,0 +1,129 @@
...
+		Assert.assertEquals("Minutes",
+				TimeUtils.parsePeriodString("60m").getPeriodType().getName());
+		Assert.assertEquals("Months",
+				TimeUtils.parsePeriodString("12M").getPeriodType().getName());
+		Assert.assertEquals("Seconds",
+				TimeUtils.parsePeriodString("120s").getPeriodType().getName());
+	}
+
+	@Test
+	public void testCreatePeriodString() {
```
Comment (petersear) [inline az-core/src/test/java/azkaban/utils/TimeUtilsTest.java]: Order the asserts logically.

File: azkaban-common/src/test/java/azkaban/utils/StringUtilsTest.java [hunk 19]
```diff
@@ -1,80 +1,138 @@
...
 import org.junit.Test;
 
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collection;
+
 public class StringUtilsTest {
 
   private static final String chromeOnMac =
-      "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_9_4) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/44.0.2403.155 Safari/537.36";
```
Comment (petersear) [inline azkaban-common/src/test/java/azkaban/utils/StringUtilsTest.java]: There seem to be a lot of whitespace changes here (spaces have become tabs?)

File: azkaban-common/src/test/java/azkaban/utils/StringUtilsTest.java [hunk 20]
```diff
@@ -1,80 +1,138 @@
...
-      Assert.assertTrue(StringUtils.isFromBrowser("awesome " + name + " is"));
-    }
+	for (final String name : BROWSER_NAMES) {
+	  Assert.assertTrue(StringUtils.isFromBrowser("awesome " + name + " is"));
+	}
+  }
+
+  @Test
+  public void testShellQuote() {
+	Assert.assertEquals("!\\!!", StringUtils.shellQuote("!", '!'));
```
Comment (petersear) [inline azkaban-common/src/test/java/azkaban/utils/StringUtilsTest.java]: You could use more meaningful inputs eg. a readable string.

File: azkaban-common/src/test/java/azkaban/utils/StringUtilsTest.java [hunk 21]
```diff
@@ -1,80 +1,138 @@
...
+  public void testJoin() {
+	final Collection<String> list = new ArrayList<>();
+	list.add("!!");
+	list.add("??");
+
+	Assert.assertEquals("!", StringUtils.join(
+			new ArrayList<>(Arrays.asList("")), "!"));
+	Assert.assertEquals("", StringUtils.join(
+			new ArrayList<>(), null));
+	Assert.assertEquals("!!$$$??$$$", StringUtils.join(
```
Comment (petersear) [inline azkaban-common/src/test/java/azkaban/utils/StringUtilsTest.java]: Could use more meaningful inputs (optional)

File: build.gradle [hunk 22]
```diff
@@ -125,6 +125,7 @@ ext.deps = [
         jacksonDatabind     : 'com.fasterxml.jackson.core:jackson-databind:2.7.4',
         jasypt              : 'org.jasypt:jasypt:1.9.2',
         jacksonCore         : 'com.fasterxml.jackson.core:jackson-core:2.9.2'
+
```
Comment (petersear) [inline build.gradle]: Should this change be here?
If you are not modifying any lines in the file, I wouldn't expect to see a blank line added.

File: azkaban-common/src/test/java/azkaban/utils/StringUtilsTest.java (hunk 13 above)
Comment (petersear) [reply azkaban-common/src/test/java/azkaban/utils/StringUtilsTest.java]: I just tried the suggestions above and the tests passed. Might be worth another try?

**Review by petersear — CHANGES_REQUESTED** [review]
It looks like tabs are not used in these test classes. It would be worth checking that all indents consist of the same number of spaces as the existing files rather than tabs.

File: azkaban-common/src/test/java/azkaban/utils/StringUtilsTest.java (hunk 13 above)
Comment (shafi-1) [reply azkaban-common/src/test/java/azkaban/utils/StringUtilsTest.java]: it works, I must have missed the *

File: az-core/src/test/java/azkaban/utils/TimeUtilsTest.java (hunk 14 above)
Comment (shafi-1) [reply az-core/src/test/java/azkaban/utils/TimeUtilsTest.java]: in order and removed underscores

File: az-core/src/test/java/azkaban/utils/TimeUtilsTest.java (hunk 15 above)
Comment (shafi-1) [reply az-core/src/test/java/azkaban/utils/TimeUtilsTest.java]: ordered, tried adding a period of `0` but broke the test, the lines in that method seem to be covered entirely anyway

File: az-core/src/test/java/azkaban/utils/TimeUtilsTest.java (hunk 16 above)
Comment (shafi-1) [reply az-core/src/test/java/azkaban/utils/TimeUtilsTest.java]: now in order

File: az-core/src/test/java/azkaban/utils/TimeUtilsTest.java (hunk 17 above)
Comment (shafi-1) [reply az-core/src/test/java/azkaban/utils/TimeUtilsTest.java]: swapped

File: az-core/src/test/java/azkaban/utils/TimeUtilsTest.java (hunk 18 above)
Comment (shafi-1) [reply az-core/src/test/java/azkaban/utils/TimeUtilsTest.java]: ordered

File: azkaban-common/src/test/java/azkaban/utils/StringUtilsTest.java (hunk 19 above)
Comment (shafi-1) [reply azkaban-common/src/test/java/azkaban/utils/StringUtilsTest.java]: I think I accidentally changed the spaces to tabs, switched back to spaces

File: azkaban-common/src/test/java/azkaban/utils/StringUtilsTest.java (hunk 20 above)
Comment (shafi-1) [reply azkaban-common/src/test/java/azkaban/utils/StringUtilsTest.java]: tried to add more meaningful inputs

File: azkaban-common/src/test/java/azkaban/utils/StringUtilsTest.java (hunk 21 above)
Comment (shafi-1) [reply azkaban-common/src/test/java/azkaban/utils/StringUtilsTest.java]: same here

File: build.gradle (hunk 22 above)
Comment (shafi-1) [reply build.gradle]: removed the blank line

File: az-core/src/test/java/azkaban/utils/TimeUtilsTest.java (hunk 15 above)
Comment (petersear) [reply az-core/src/test/java/azkaban/utils/TimeUtilsTest.java]: My mistake, that should have been "0s" which will skip all of the `if`'s to go straight to the `return`. Maybe that isn't that useful though.

File: az-core/src/test/java/azkaban/utils/TimeUtilsTest.java (hunk 8 above)
Comment (petersear) [reply az-core/src/test/java/azkaban/utils/TimeUtilsTest.java]: Still need a `carriage return` at the end of the file as mentioned by @Braavos96

File: azkaban-common/src/test/java/azkaban/utils/StringUtilsTest.java [hunk 23]
```diff
@@ -77,4 +97,47 @@ public void containsBrowserName() {
...
+			new ArrayList<>(), null));
+	Assert.assertEquals("!!$$$??", StringUtils.join2(
+			new ArrayList<>(Arrays.asList("!!", "??")), "$$$"));
+  }
+
+  @Test
+  public void testIsFromBrowser() {
+	Assert.assertFalse(StringUtils.isFromBrowser(null));
+	Assert.assertFalse(StringUtils.isFromBrowser("foo"));
+	Assert.assertTrue(StringUtils.isFromBrowser("*Gecko"));
```
Comment (petersear) [inline azkaban-common/src/test/java/azkaban/utils/StringUtilsTest.java]: Insert a blank line between the different assert types (`False` and `True`)

File: az-core/src/test/java/azkaban/utils/TimeUtilsTest.java [hunk 24]
```diff
@@ -0,0 +1,130 @@
+/*
+ * Copyright 2017 LinkedIn Corp.
```
Comment (petersear) [inline az-core/src/test/java/azkaban/utils/TimeUtilsTest.java:2]: They might want this updated to `2019`. On the other hand, I had feedback not to change the year so maybe just be aware.

**Review by petersear — CHANGES_REQUESTED** [review]
There are still some tabs in both files although it's looking better :-)
The style seems to be 2-space indents in the existing (and other) test class so worth maybe changing your IDE settings accordingly and matching the project.

File: azkaban-common/src/test/java/azkaban/utils/StringUtilsTest.java (hunk 23 above)
Comment (shafi-1) [reply azkaban-common/src/test/java/azkaban/utils/StringUtilsTest.java]: inserted

File: az-core/src/test/java/azkaban/utils/TimeUtilsTest.java (hunk 24 above)
Comment (shafi-1) [reply az-core/src/test/java/azkaban/utils/TimeUtilsTest.java:2]: noted

File: az-core/src/test/java/azkaban/utils/TimeUtilsTest.java [hunk 25]
```diff
@@ -0,0 +1,131 @@
...
+		Assert.assertEquals("1m 1s",
+				TimeUtils.formatDuration(-65793L, -4609L));
+		Assert.assertEquals("1h 14m 1s",
+				TimeUtils.formatDuration(-4462057L, -20994L));
+		Assert.assertEquals("1290d 15h 0m",
+				TimeUtils.formatDuration(-833737520625L, -722227520546L));
+	}
+
+	@Test
+	public void testParsePeriodStringThrowsException() {
```
Comment (petersear) [inline az-core/src/test/java/azkaban/utils/TimeUtilsTest.java]: There is another test for an exception thrown by parsePeriodString() below testParsePeriodString(), which is the right location for it, so this one could be removed I think?

File: az-core/src/test/java/azkaban/utils/TimeUtilsTest.java [hunk 26]
```diff
@@ -0,0 +1,131 @@
...
+ */
+
+package azkaban.utils;
+
+import org.junit.Assert;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.rules.ExpectedException;
+
+public class TimeUtilsTest {
```
Comment (petersear) [inline az-core/src/test/java/azkaban/utils/TimeUtilsTest.java:24]: This whole (new) test class is indented using tabs but should be indented using 2 spaces to match other existing test files.

File: az-core/src/test/java/azkaban/utils/TimeUtilsTest.java (hunk 8 above)
Comment (petersear) [reply az-core/src/test/java/azkaban/utils/TimeUtilsTest.java]: Still needs resolving.

File: azkaban-common/src/test/java/azkaban/utils/StringUtilsTest.java [hunk 27]
```diff
@@ -1,8 +1,28 @@
...
 package azkaban.utils;
 
 import org.junit.Assert;
 import org.junit.Test;
 
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collection;
+
 public class StringUtilsTest {
```
Comment (petersear) [inline azkaban-common/src/test/java/azkaban/utils/StringUtilsTest.java:26]: The existing tests in this class use 2 spaces for indents (and 4 for a continuation line)
The new tests use an 8-space tab which is equivalent to indenting by 6 spaces relative to the previous line.
This should be changed so that the style of the added tests matches that of the existing tests.

**Review by petersear — CHANGES_REQUESTED** [review]

File: az-core/src/test/java/azkaban/utils/TimeUtilsTest.java (hunk 8 above)
Comment (shafi-1) [reply az-core/src/test/java/azkaban/utils/TimeUtilsTest.java]: i added a blank line at the end of the file

File: az-core/src/test/java/azkaban/utils/TimeUtilsTest.java (hunk 25 above)
Comment (shafi-1) [reply az-core/src/test/java/azkaban/utils/TimeUtilsTest.java]: removed the extra test

File: az-core/src/test/java/azkaban/utils/TimeUtilsTest.java (hunk 26 above)
Comment (shafi-1) [reply az-core/src/test/java/azkaban/utils/TimeUtilsTest.java:24]: managed to change

File: azkaban-common/src/test/java/azkaban/utils/StringUtilsTest.java (hunk 27 above)
Comment (shafi-1) [reply azkaban-common/src/test/java/azkaban/utils/StringUtilsTest.java:26]: changed to match the style

**Review by petersear — APPROVED** [review]

**Review by JohnLBergqvist — APPROVED** [review]

**Review by paulkennethkent — APPROVED** [review]
